### ISSUE TITLE
Loading images with non-inferred labels

### ISSUE THREAD

**Issue by Marco-Christiani** [tensorflow/tensorflow#44752]
**System information**
- OS: macOs Catalina
- Device: MacBook Pro 2017
- GPU: Radeon Pro 555 2 GB & Intel HD Graphics 630 1536 MB
- Memory: 16 GB 2133 MHz LPDDR3
- Python Version: 3.7.9

`tf.version.GIT_VERSION: v2.3.0-54-gfcc4b966f1`
`tf.version.VERSION: 2.3.1`

**Current behavior**
`image_dataset_from_directory()` seemingly ignores `labels` argument

**Expected behavior**
When running `train_ds = image_dataset_from_directory(my_dir, labels=my_label_list)` with `my_label_list` containing integer labels with the exact number of elements as the number of images in `my_dir` (and in alphanumerically correct order with relation to the image filenames) I would expect `train_ds.class_names` to consist of the integer labels passed.

From the [documentation](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/image_dataset_from_directory) I would expect the directory structure to be ignored since I have supplied the `labels` parameter. However, images are only loaded when I nest them in a sub-directory (in this example named `images/`). This sub-directory is then assigned as the sole class label.

```python
>>> print(train_ds.class_names)
['images']
```

**Standalone code to reproduce the issue**

Some sample images [here](https://drive.google.com/file/d/1msGXyy-8aDyeqPmsRCfeWiS3iAbzLLMm/view?usp=sharing) and follow this directory structure:
```
test_dir/
└───images/
```
Then run the following example:
```python
image_dir = '/foo/bar/test_dir/'
label_list = [1,2,3,1,1]
train_ds = tf.keras.preprocessing.image_dataset_from_directory(
    image_dir,
    labels=label_list,
    validation_split=0.2,
    subset='training',
    shuffle=False,
)
print(train_ds.class_names)
```

**Comment by ravikyram** [comment]
I have tried in colab with TF version 2.3, nightly version and was able to reproduce the issue. Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/e4b12d66766306aae5502909fb776a67/untitled511.ipynb).Thanks!

**Comment by bhack** [comment]
The `class_names` it is temporary fixed at https://github.com/tensorflow/tensorflow/pull/44769 but we could talk in that PR about the behavior that we want for directory/subdirectory walking of `image_paths`.

**Comment by bhack** [comment]
The PR fixed another issue

**Comment by sachinprasadhs** [comment]
Was able to reproduce your issue in Tf Nightly 2.6.0-dev20210530, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/a9e7c6ba293a3b0d2d949a8c45d47a1f/44752.ipynb). Thanks!

**Comment by sachinprasadhs** [comment]
`class_names` Only valid if "labels" is "inferred". This is the explicit list of class names (must match names of subdirectories). Used to control the order of the classes (otherwise alphanumerical order is used).

Your `train_ds` yields a tuple (images, labels), where images has shape `(batch_size, image_size[0], image_size[1], num_channels)`, and `labels` follows the format described below.

To get the class names you can get it from the below lines of code.

```
for i in train_ds.take(5):
  print(i[1])
```

`tf.Tensor([1 2 3 1], shape=(4,), dtype=int32)`

Attaching the gist [here](https://colab.sandbox.google.com/gist/sachinprasadhs/a3bb655ecbf1f21c6acf70fd41a6439a/44752.ipynb) for reference. Thanks!
